### PR TITLE
[Dynamo] Save global autocast state to restore on graph break

### LINKED ISSUE #102414: [Dynamo] graph break under autocast context manager inlining failure caused NE gap
### 🐛 Describe the bug

This is a bug from Meta internal models, where we observed model's NE gap between PT2 and PT1, and the root cause is similar to https://github.com/pytorch/pytorch/pull/101201 where the global state is changed when inlining failed, but we don't restore it correctly.

Repro:
```
import torch
import torch._dynamo
import logging

# torch._logging.set_logs(dynamo=logging.DEBUG, aot=logging.DEBUG, inductor=logging.DEBUG)

def gn(x, y):
    with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
        z = torch.mm(x, y)
        torch._dynamo.graph_break()
        return torch.sin(z)


def fn(x, y):
    z = torch.mm(x, y)
    z = z + gn(x, y)
    return z


x = torch.rand(3, 3, device="cuda")
y = torch.rand(3, 3, device="cuda")

print(fn(x, y))
opt_fn = torch.compile(backend="eager")(fn)
print(opt_fn(x, y))
```
Output:
```
tensor([[1.3335, 1.5669, 1.3017],
        [0.6343, 2.1670, 1.0982],
        [0.6622, 1.2455, 1.1671]], device='cuda:0')
tensor([[1.3359, 1.5703, 1.2969],
        [0.6328, 2.1719, 1.0938],
        [0.6641, 1.2500, 1.1719]], device='cuda:0', dtype=torch.bfloat16)
```

### Versions

N/A

cc @ezyang @msaroufim @wconstab @ngimel @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #102414

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: test/dynamo/test_ctx_manager.py [hunk 1]
```diff
@@ -697,6 +697,33 @@ def fn(x, y, z):
...
+        res = opt_fn(x, y)
+        self.assertEqual(ref, res)
+
+    def test_graph_break_inlining_autocast(self):
+        for device in ["cuda", "cpu"]:
+            if device == "cuda" and not (
+                torch.cuda.is_available() and torch.cuda.is_bf16_supported()
+            ):
+                continue
+            self._graph_break_inlining_autocast_test_helper(device)
```
Comment (yf225) [inline test/dynamo/test_ctx_manager.py:725]: Curious: with current impl, we will only run `_graph_break_inlining_autocast_test_helper` if CUDA is available and BF16 is supported. Do we always have CI machine that meets this requirement?

**Comment by anijain2305** [comment]
There is also a `autocast_cache_enabled` thing. I am not sure what is it used for, but maybe we need to save-restore that well. cc @eellison

File: test/dynamo/test_ctx_manager.py (hunk 1 above)
Comment (yanboliang) [reply test/dynamo/test_ctx_manager.py:726]: Yes, we have many CI machines with CUDA and BF16 enabled. This check is used to skip CUDA/BF16 test cases when running on CPU only CI machines.

**Comment by yanboliang** [comment]
> There is also a autocast_cache_enabled thing. I am not sure what is it used for, but maybe we need to save-restore that well.

Added

**Review by yf225 — APPROVED** [review]
Thanks Yanbo!

**Pull request closed without merging**